sourcegit/config.py

```diff
@@ -28,6 +34,88 @@ def get_default_map_from_file() -> Optional[dict]:
 
 @lru_cache()
 def get_context_settings() -> dict:
-    return dict(help_option_names=['-h', '--help'],
-                auto_envvar_prefix='SOURCE_GIT',
-                default_map=get_default_map_from_file())
+    return dict(
+        help_option_names=["-h", "--help"],
+        auto_envvar_prefix="SOURCE_GIT",
+        default_map=get_default_map_from_file(),
+    )
+
+
+class TriggerType(Enum):
+    release = 1
+    pull_request = 2
+    branch_commit = 3
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class JobConfig:
+    trigger: TriggerType
+    release_to: List[str]
+    metadata: dict
+
+    @classmethod
+    def get_from_dict(cls, raw_dict: dict, validate=True) -> JobConfig:
+        if validate and not JobConfig.is_dict_valid(raw_dict):
+            raise Exception(f"Job config not valid.")
+
+        trigger_raw, release_to, metadata = exclude_from_dict(
+            raw_dict, "trigger", "release_to"
+        )
+        return JobConfig(
+            trigger=TriggerType[trigger_raw], release_to=release_to, metadata=metadata
+        )
+
+    @classmethod
+    def is_dict_valid(cls, raw_dict: dict) -> bool:
+        return Draft4Validator(JOB_CONFIG_SCHEMA).is_valid(raw_dict)
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class PackageConfig:
+    specfile: str
+    synced_files: List[str]
+    jobs: List[JobConfig]
+    hooks: Optional[Dict[str, str]]  # action_name: script
+    metadata: dict
+
+    @classmethod
+    def get_from_dict(cls, raw_dict: dict, validate=True) -> PackageConfig:
+        if validate and not PackageConfig.is_dict_valid(raw_dict):
+            raise Exception("Package config not valid.")
+
+        spec_file, synced_files, raw_jobs, hooks, metadata = exclude_from_dict(
+            raw_dict, "specfile", "synced_files", "jobs", "hooks"
+        )
+
+        return PackageConfig(
+            specfile=spec_file,
+            synced_files=synced_files,
+            jobs=[JobConfig.get_from_dict(raw_job, validate=False) for raw_job in raw_jobs],
+            hooks=hooks,
+            metadata=metadata,
+        )
+
+    @classmethod
+    def is_dict_valid(cls, raw_dict: dict) -> bool:
+        return Draft4Validator(PACKAGE_CONFIG_SCHEMA).is_valid(raw_dict)
+
+
+JOB_CONFIG_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "trigger": {"enum": ["release", "pull_request", "branch_commit"]},
```

Comment:
How about:
```
["git_tag", "github_release", "pull_request"]
```
I would add branch_commit later as an increment.

I'm torn on the release/github_release choice. Release is better since we won't be locked to a platform, but we might need some more data on where the release is coming from.

Reply:
OK, I like `release`, since it's more general.

So: `git_tag`, `release` and `pull_request` for now.

Reply:
fixed